File: backend/rag/query_intent.py
"""Deterministic query understanding for retrieval."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum

from .relevance import QUERY_EXPANSIONS, _terms
from .mechanism import extract_phrase_concepts, ignored_stopwords

AUTHOR_RE = re.compile(r"\bauthor:([A-Za-z][\w'’-]*)", re.IGNORECASE)
YEAR_RE = re.compile(r"\byear:(\d{4})(?:-(\d{4}))?", re.IGNORECASE)
SECTION_RE = re.compile(r'\bsection:(?:"([^"]+)"|([^\s]+))', re.IGNORECASE)
CATEGORY_RE = re.compile(r"\bcategory:([A-Za-z][\w-]*)", re.IGNORECASE)
TOPICAL_RE = re.compile(r"\btopical:(true|false|yes|no|1|0)", re.IGNORECASE)
COUNT_RE = re.compile(r"\bcount:(\d{1,2})", re.IGNORECASE)
OPPONENT_PREFIX_RE = re.compile(
    r"^\s*(?:opponent|they|other team)\s+(?:says?|argues?|claims?)\s+", re.IGNORECASE
)
CARD_CITATION_RE = re.compile(
    r"^\s*(?P<author>[A-Za-z][A-Za-z'’-]{1,})\s+"
    r"(?P<year>\d{4}|[‘'’]\d{2}|\d{2})\s*$"
)
AUTHOR_ONLY_RE = re.compile(r"^\s*[A-Za-z][A-Za-z'’-]{2,}\s*$")
SECTION_LOOKUP_RE = re.compile(r"^\s*(?:AT:|OV\b|Overview\b).+", re.IGNORECASE)


class SearchMode(str, Enum):
    ARGUMENT = "argument"
    AUTHOR = "author"
    CITATION = "citation"
    SECTION = "section"
    GENERAL = "general"


@dataclass(frozen=True)
class QueryIntent:
    raw_query: str
    mode: str = "search"
    search_mode: SearchMode = SearchMode.GENERAL
    search_text: str = ""
    opponent_claim: str | None = None
    concepts: list[str] = field(default_factory=list)
    phrase_concepts: list[str] = field(default_factory=list)
    ignored_stopwords: list[str] = field(default_factory=list)
    author_filter: str | None = None
    year_min: int | None = None
    year_max: int | None = None
    section_filter: str | None = None
    category_filter: str | None = None
    topical_filter: bool | None = None
    requested_count: int | None = None


def parse_query_intent(query: str, mode: str = "search") -> QueryIntent:
    """Parse simple exact filters and debate concepts without calling an LLM."""
    author_filter = _first_group(AUTHOR_RE, query)
    section_filter = _section_filter(query)
    category_filter = _first_group(CATEGORY_RE, query)
    topical_filter = _topical_filter(query)
    requested_count = _count_filter(query)
    year_min, year_max = _year_filter(query)

    search_text = _strip_filters(query)
    opponent_claim = _opponent_claim(search_text)
    citation_match = CARD_CITATION_RE.match(search_text)
    if citation_match and not author_filter:
        author_filter = citation_match.group("author")
        year_min = year_max = _normalize_year(citation_match.group("year"))
    if SECTION_LOOKUP_RE.match(search_text) and not section_filter:
        section_filter = search_text
    if AUTHOR_ONLY_RE.match(search_text) and not author_filter:
        author_filter = search_text

    search_mode = _search_mode(
        search_text=search_text,
        opponent_claim=opponent_claim,
        author_filter=author_filter,
        year_min=year_min,
        section_filter=section_filter,
        citation_match=bool(citation_match),
    )
    concept_text = opponent_claim or search_text
    concepts = _concepts(concept_text)
    phrase_concepts = sorted(extract_phrase_concepts(concept_text))
    ignored = sorted(ignored_stopwords(query))

    return QueryIntent(
        raw_query=query,
        mode=mode,
        search_mode=search_mode,
        search_text=search_text,
        opponent_claim=opponent_claim,
        concepts=concepts,
        phrase_concepts=phrase_concepts,
        ignored_stopwords=ignored,
        author_filter=author_filter,
        year_min=year_min,
        year_max=year_max,
        section_filter=section_filter,
        category_filter=category_filter,
        topical_filter=topical_filter,
        requested_count=requested_count,
    )


def _strip_filters(query: str) -> str:
    text = query
    for pattern in (AUTHOR_RE, YEAR_RE, SECTION_RE, CATEGORY_RE, TOPICAL_RE, COUNT_RE):
        text = pattern.sub("", text)
    return " ".join(text.split())


def _opponent_claim(text: str) -> str | None:
    claim = OPPONENT_PREFIX_RE.sub("", text).strip(" .")
    if claim != text.strip(" .") and claim:
        return claim
    return None


def _concepts(text: str) -> list[str]:
    terms = _terms(text)
    expanded = set(terms)
    for term in terms:
        expanded.update(QUERY_EXPANSIONS.get(term, set()))
    return sorted(expanded)


def _first_group(pattern: re.Pattern[str], text: str) -> str | None:
    match = pattern.search(text)
    return match.group(1) if match else None


def _section_filter(text: str) -> str | None:
    match = SECTION_RE.search(text)
    if not match:
        return None
    return match.group(1) or match.group(2)


def _year_filter(text: str) -> tuple[int | None, int | None]:
    match = YEAR_RE.search(text)
    if not match:
        return None, None
    start = int(match.group(1))
    end = int(match.group(2) or match.group(1))
    return start, end


def _normalize_year(value: str) -> int:
    digits = value.strip("‘'’")
    if len(digits) == 4:
        return int(digits)
    year = int(digits)
    return 2000 + year if year < 70 else 1900 + year


def _search_mode(
    *,
    search_text: str,
    opponent_claim: str | None,
    author_filter: str | None,
    year_min: int | None,
    section_filter: str | None,
    citation_match: bool,
) -> SearchMode:
    if opponent_claim:
        return SearchMode.ARGUMENT
    if section_filter and SECTION_LOOKUP_RE.match(section_filter):
        return SearchMode.SECTION
    if author_filter and year_min is not None:
        return SearchMode.CITATION
    if author_filter:
        return SearchMode.AUTHOR
    if citation_match:
        return SearchMode.CITATION
    if AUTHOR_ONLY_RE.match(search_text):
        return SearchMode.AUTHOR
    return SearchMode.GENERAL


def _topical_filter(text: str) -> bool | None:
    match = TOPICAL_RE.search(text)
    if not match:
        return None
    return match.group(1).lower() in {"true", "yes", "1"}


def _count_filter(text: str) -> int | None:
    match = COUNT_RE.search(text)
    if not match:
        return None
    return max(1, min(int(match.group(1)), 50))

File: app/backend/hybrid/query_intent.cpp
#include "query_intent.hpp"

#include "mechanism.hpp"
#include "relevance.hpp"

#include <algorithm>
#include <cctype>
#include <regex>
#include <set>
#include <sstream>

namespace sekret::hybrid {
namespace {

std::string trim(const std::string& text) {
    auto begin = text.begin();
    while (begin != text.end() && std::isspace(static_cast<unsigned char>(*begin))) {
        ++begin;
    }
    auto end = text.end();
    while (end != begin && std::isspace(static_cast<unsigned char>(*(end - 1)))) {
        --end;
    }
    return std::string(begin, end);
}

std::string trim_dots_spaces(const std::string& text) {
    auto value = trim(text);
    while (!value.empty() && (value.back() == '.' || std::isspace(static_cast<unsigned char>(value.back())))) {
        value.pop_back();
    }
    return trim(value);
}

std::string squeeze_spaces(const std::string& text) {
    std::ostringstream output;
    bool previous_space = true;
    for (const unsigned char character : text) {
        if (std::isspace(character)) {
            if (!previous_space) {
                output << ' ';
            }
            previous_space = true;
            continue;
        }
        output << static_cast<char>(character);
        previous_space = false;
    }
    return trim(output.str());
}

std::string regex_strip(std::string text, const std::regex& pattern) {
    return std::regex_replace(text, pattern, "");
}

std::optional<std::string> first_group(const std::regex& pattern, const std::string& text) {
    std::smatch match;
    if (!std::regex_search(text, match, pattern) || match.size() < 2) {
        return std::nullopt;
    }
    return match[1].str();
}

std::optional<std::string> section_filter(const std::string& text) {
    static const std::regex pattern(R"REGEX(\bsection:(?:"([^"]+)"|([^\s]+)))REGEX", std::regex::icase);
    std::smatch match;
    if (!std::regex_search(text, match, pattern)) {
        return std::nullopt;
    }
    if (match[1].matched) {
        return match[1].str();
    }
    return match[2].str();
}

std::pair<std::optional<int>, std::optional<int>> year_filter(const std::string& text) {
    static const std::regex pattern(R"(\byear:(\d{4})(?:-(\d{4}))?)", std::regex::icase);
    std::smatch match;
    if (!std::regex_search(text, match, pattern)) {
        return {std::nullopt, std::nullopt};
    }
    const int start = std::stoi(match[1].str());
    const int end = match[2].matched ? std::stoi(match[2].str()) : start;
    return {start, end};
}

std::optional<bool> topical_filter(const std::string& text) {
    static const std::regex pattern(R"(\btopical:(true|false|yes|no|1|0))", std::regex::icase);
    std::smatch match;
    if (!std::regex_search(text, match, pattern)) {
        return std::nullopt;
    }
    auto value = match[1].str();
    std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    });
    return value == "true" || value == "yes" || value == "1";
}

std::optional<int> count_filter(const std::string& text) {
    static const std::regex pattern(R"(\bcount:(\d{1,2}))", std::regex::icase);
    std::smatch match;
    if (!std::regex_search(text, match, pattern)) {
        return std::nullopt;
    }
    return std::max(1, std::min(std::stoi(match[1].str()), 50));
}

std::string strip_filters(const std::string& query) {
    std::string text = query;
    for (const auto& pattern : {
        std::regex(R"(\bauthor:([A-Za-z][\w'’-]*))", std::regex::icase),
        std::regex(R"(\byear:(\d{4})(?:-(\d{4}))?)", std::regex::icase),
        std::regex(R"REGEX(\bsection:(?:"([^"]+)"|([^\s]+)))REGEX", std::regex::icase),
        std::regex(R"(\bcategory:([A-Za-z][\w-]*))", std::regex::icase),
        std::regex(R"(\btopical:(true|false|yes|no|1|0))", std::regex::icase),
        std::regex(R"(\bcount:(\d{1,2}))", std::regex::icase),
    }) {
        text = regex_strip(text, pattern);
    }
    return squeeze_spaces(text);
}

std::optional<std::string> opponent_claim(const std::string& text) {
    static const std::regex pattern(
        R"(^\s*(?:(?:the\s+)?opposing\s+team|opponent|they|other\s+team)\s+(?:says?|argues?|claims?)\s+(?:that\s+)?)",
        std::regex::icase
    );
    const auto stripped = std::regex_replace(text, pattern, "");
    if (trim_dots_spaces(stripped) != trim_dots_spaces(text) && !trim_dots_spaces(stripped).empty()) {
        return trim_dots_spaces(stripped);
    }
    return std::nullopt;
}

int normalize_year(std::string value) {
    value.erase(
        std::remove_if(value.begin(), value.end(), [](unsigned char character) {
            return std::isdigit(character) == 0;
        }),
        value.end()
    );
    if (value.size() == 4) {
        return std::stoi(value);
    }
    const int year = std::stoi(value);
    return year < 70 ? 2000 + year : 1900 + year;
}

SearchMode detect_search_mode(
    const std::string& search_text,
    const std::optional<std::string>& claim,
    const std::optional<std::string>& author,
    const std::optional<int>& year_min,
    const std::optional<std::string>& section,
    bool citation_match
) {
    static const std::regex section_lookup(R"(^\s*(?:AT:|OV\b|Overview\b).+)", std::regex::icase);
    static const std::regex author_only(R"(^\s*[A-Za-z][A-Za-z'’-]{2,}\s*$)");
    if (claim.has_value()) {
        return SearchMode::Argument;
    }
    if (section.has_value() && std::regex_match(*section, section_lookup)) {
        return SearchMode::Section;
    }
    if (author.has_value() && year_min.has_value()) {
        return SearchMode::Citation;
    }
    if (author.has_value()) {
        return SearchMode::Author;
    }
    if (citation_match) {
        return SearchMode::Citation;
    }
    if (std::regex_match(search_text, author_only)) {
        return SearchMode::Author;
    }
    return SearchMode::General;
}

std::vector<std::string> sorted_vector(const std::set<std::string>& values) {
    return {values.begin(), values.end()};
}

std::vector<std::string> concepts_for_text(const std::string& text) {
    auto values = terms(text);
    auto phrases = extract_phrase_concepts(text);
    values.insert(phrases.begin(), phrases.end());
    return sorted_vector(values);
}

} // namespace

QueryIntent parse_query_intent(const std::string& query, const std::string& mode) {
    static const std::regex author_re(R"(\bauthor:([A-Za-z][\w'’-]*))", std::regex::icase);
    static const std::regex category_re(R"(\bcategory:([A-Za-z][\w-]*))", std::regex::icase);
    static const std::regex card_citation_re(R"(^\s*([A-Za-z][A-Za-z'’-]{1,})\s+(\d{4}|[‘'’]\d{2}|\d{2})\s*$)");
    static const std::regex author_only_re(R"(^\s*[A-Za-z][A-Za-z'’-]{2,}\s*$)");
    static const std::regex section_lookup_re(R"(^\s*(?:AT:|OV\b|Overview\b).+)", std::regex::icase);

    QueryIntent intent;
    intent.raw_query = query;
    intent.mode = mode;
    intent.author_filter = first_group(author_re, query);
    intent.section_filter = section_filter(query);
    intent.category_filter = first_group(category_re, query);
    intent.topical_filter = topical_filter(query);
    intent.requested_count = count_filter(query);
    auto years = year_filter(query);
    intent.year_min = years.first;
    intent.year_max = years.second;
    intent.search_text = strip_filters(query);
    intent.opponent_claim = opponent_claim(intent.search_text);

    std::smatch citation_match;
    const bool is_citation = std::regex_match(intent.search_text, citation_match, card_citation_re);
    if (is_citation && !intent.author_filter.has_value()) {
        intent.author_filter = citation_match[1].str();
        const int normalized = normalize_year(citation_match[2].str());
        intent.year_min = normalized;
        intent.year_max = normalized;
    }
    if (std::regex_match(intent.search_text, section_lookup_re) && !intent.section_filter.has_value()) {
        intent.section_filter = intent.search_text;
    }
    if (std::regex_match(intent.search_text, author_only_re) && !intent.author_filter.has_value()) {
        intent.author_filter = intent.search_text;
    }

    intent.search_mode = detect_search_mode(
        intent.search_text,
        intent.opponent_claim,
        intent.author_filter,
        intent.year_min,
        intent.section_filter,
        is_citation
    );
    const auto concept_text = intent.opponent_claim.value_or(intent.search_text);
    intent.concepts = concepts_for_text(concept_text);
    intent.phrase_concepts = sorted_vector(extract_phrase_concepts(concept_text));
    intent.ignored_stopwords = sorted_vector(ignored_stopwords(query));
    return intent;
}

std::string search_mode_name(SearchMode mode) {
    switch (mode) {
        case SearchMode::Argument:
            return "argument";
        case SearchMode::Author:
            return "author";
        case SearchMode::Citation:
            return "citation";
        case SearchMode::Section:
            return "section";
        case SearchMode::General:
            return "general";
    }
    return "general";
}

std::string retrieval_text(const QueryIntent& intent) {
    if (intent.opponent_claim.has_value()) {
        return *intent.opponent_claim;
    }
    if (!intent.search_text.empty()) {
        return intent.search_text;
    }
    return intent.raw_query;
}

} // namespace sekret::hybrid
